advertise/addr.go

```diff
@@ -13,91 +14,152 @@ import (
 // local network traffic for Unix-like platforms.
 var DefaultInterfaces = []string{"eth0", "en0"}
 
-// FirstAddress returns the first IPv4 address from the given interface names.
-// Addresses used for APIPA will be ignored if possible.
-func FirstAddress(interfaces []string) (net.IP, error) {
+// FirstAddress returns the "best" IP address from the given interface names.
+// "best" is defined as follow in decreasing order:
+// - IPv4 valid and not link-local unicast
+// - IPv6 valid and not link-local unicast
+// - IPv4 valid and link-local unicast
+// - IPv6 valid and link-local unicast
+// If none of the above are found, an invalid address is returned.
+// Loopback addresses are never selected.
+// If no interfaces are provided, all of the system's network interfaces will retrieved via net.Interfaces
+// and used to find the "best" ip address.
+func FirstAddress(interfaces []string) (string, error) {
+	return firstAddress(interfaces, getInterfaceAddresses, net.Interfaces)
+}
```

Comment:
Is there a reason this doesn't return a `net.IP` anymore?

Reply:
dskit returns a string now and in our code we just use the string of the net.IP so I decided to follow dskit implementation. Do you see a particular benefit to return the net.IP instead of the string?

Reply:
The main reason is because this change would technically be a breaking change to the API.

The secondary reason is because there's tons of [methods](https://pkg.go.dev/net#IP) around IP addresses that may be useful to callers. With this change, a user would have to parse the IP again to make use of any of those metrics.

Reply:
okay, let's return the net.IP, we already differentiate from dskit in other ways anyway

Reply:
ah actually I forgot, the code is not using net.IP anymore but netip.Addr instead

Reply:
@rfratto I made a new version where I return the netip.Addr. This will still be a breaking change and we can't return `nil` anymore, but the caller will have the methods around netip.Addr